### PR TITLE
Update INA228.cpp

### PR DESCRIPTION
**Pull request by markliquid1**
updated getShuntVoltage() to allow reading of negative currents

### CODE REVIEW

**Comment by RobTillaart** [comment]
Please have a look at PR #7, there is similar work in progress

**Comment by RobTillaart** [comment]
If I understand correctly this should allow the small negative range over the bus (and thus negative current for shunt)?

Datasheet 7.3.1
_The supported common-mode voltage range at the input pins is –0.3 V to +85
V, which makes the device well suited for both high-side and low-side current measurements._

Next weekend I hope to have time to get hands on with hardware.

File: INA228.cpp [hunk 1]
```diff
@@ -81,16 +81,31 @@ float INA228::getBusVoltage()
...
+//  return value * 312.5e-9;
+//}
+
+//Proposed new function which allows the reading of both positive and negative currents
+float INA228::getShuntVoltage() {
   uint16_t config = _readRegister(INA228_CONFIG, 2);
-  if (config & 0x0008) return value * 78.125e-9;
-  return value * 312.5e-9;
+  float scale = 312.5;
+  if (config & 0x0008) {
```
Comment (RobTillaart) [inline INA228.cpp:99]: mask must be 0x0010 as the ADCRANGE is bit 4  or  (1 << 4)

File: INA228.cpp [hunk 2]
```diff
@@ -81,16 +81,31 @@ float INA228::getBusVoltage()
...
+    scale = 78.125;
+  }
+  int32_t value = _readRegister(INA228_SHUNT_VOLTAGE, 3);
+  if (value & 0x800000)
```
Comment (RobTillaart) [inline INA228.cpp:103]: value is not shifted yet so the sign bit test should test 0x80000000.

File: INA228.cpp [hunk 3]
```diff
@@ -81,16 +81,31 @@ float INA228::getBusVoltage()
...
+    value |= 0xFF000000;   
+  return (float)value / 16.0 * scale / 1000000.0 / 1000;
```
Comment (RobTillaart) [inline INA228.cpp:105]: bit unclear what happens here. I will post a refactored version including fixes for the above.

**Comment by RobTillaart** [comment]
Have checked the data sheet and your code is an improvement however with minor points see above.

```cpp
float INA228::getShuntVoltage() 
{
  //  datasheet P25
  float LSB = 312.5e-9;  //  312.5 nV
  uint16_t config = _readRegister(INA228_CONFIG, 2);
  if (config & 0x0010)   //  ADCRANGE
  {
    LSB = 78.125e-9;     //  78.125 nV
  }

  //  remove reserved bits.
  int32_t value = _readRegister(INA228_SHUNT_VOLTAGE, 3) >> 4;
  if (value & 0x00800000) 
  {
    value |= 0xFF000000;
  }
  return value * LSB;
}
```

- renamed scale to LSB, 
- gave LSB the value as defined in the datasheet instead of scaling in return command
- adjusted the mask for ADCRANGE
- adjusted the mask for the SIGN bit test.

Please give it a try, thanks.

**Comment by RobTillaart** [comment]
@markliquid1 
better use this

```cpp
float INA228::getShuntVoltage() 
{
  //  datasheet P25
  float LSB = 312.5e-9;  //  312.5 nV
  if (getADCRange() == 1) 
  {
    LSB = 78.125e-9;     //  78.125 nV
  }

  //  remove reserved bits.
  int32_t value = _readRegister(INA228_SHUNT_VOLTAGE, 3) >> 4;
  //  handle negative values
  if (value & 0x00800000) 
  {
    value |= 0xFF000000;
  }
  return value * LSB;
}

**Comment by markliquid1** [comment]
Thanks, I'll give it a try today.  I'm new to programming, so mainly submitted this request as a way to point out the issue with negative current measurement.  The code I proposed was inspired by the Adafruit library for the INA228, and did seem to work, but certainly was not the cleanest way to do it, and not consistent with the rest of your library.

**Comment by RobTillaart** [comment]
Then I will include it in my current develop branch which I'm working on.
There is still work todo before the library is functional complete.

Thanks for the clarification

Will close this as a PR.
If you have more remarks, please open an issue per topic.

**Comment by markliquid1** [comment]
Ok, I wasn't yet aware of the "Issue" functionality of github, so I will use that in the future.  Thanks for all your open source work on Arduino libraries.  I have benefited greatly as I learn, including this INA228.  I look forward to your updates.

**Comment by RobTillaart** [comment]
Included getShuntVoltage() into my develop branch,

Thanks again for pointing out the negative values!

**Pull request closed without merging**